genkit-tools/common/src/eval/validate.ts

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Ajv, { ErrorObject, JSONSchemaType } from 'ajv';
+import addFormats from 'ajv-formats';
+import { getDatasetStore } from '.';
+import { RuntimeManager } from '../manager';
+import {
+  Action,
+  InferenceDatasetSchema,
+  ValidateDataRequest,
+  ValidateDataResponse,
+} from '../types';
+
+// Setup for AJV
+type JSONSchema = JSONSchemaType<any> | any;
+const ajv = new Ajv();
+addFormats(ajv);
+
+/**
+ * Validate given data against a target action. Intended to be used via the
+ * reflection API.
+ */
+export async function validateSchema(
+  manager: RuntimeManager,
+  request: ValidateDataRequest
+): Promise<ValidateDataResponse> {
+  const { dataSource, actionRef } = request;
+  const { datasetId, data } = dataSource;
+  if (!datasetId && !data) {
+    throw new Error(`Either 'data' or 'datasetId' must be provided`);
+  }
+  const targetAction = await getAction(manager, actionRef);
+  const targetSchema = targetAction?.inputSchema;
+  if (!targetAction) {
+    throw new Error(`Could not find matching action for ${actionRef}`);
+  }
+  if (!targetSchema) {
+    return { valid: true };
+  }
+
+  let errorsMap: Record<string, any> = {};
+
+  if (datasetId) {
+    const datasetStore = await getDatasetStore();
+    const dataset = await datasetStore.getDataset(datasetId);
+    if (dataset.length === 0) {
+      return { valid: true };
+    }
+    dataset.forEach((sample, index) => {
+      const response = validate(targetSchema, sample.input);
+      if (!response.valid) {
+        errorsMap[sample.testCaseId] = response.errors;
+      }
+    });
+
+    return Object.keys(errorsMap).length === 0
+      ? { valid: true }
+      : { valid: false, datasetErrors: errorsMap };
+  } else {
+    const dataset = InferenceDatasetSchema.parse(data);
+    dataset.forEach((sample, index) => {
+      const response = validate(targetSchema, sample.input);
+      if (!response.valid) {
+        errorsMap[index] = response.errors;
+      }
+    });
+    return Object.keys(errorsMap).length === 0
+      ? { valid: true }
+      : { valid: false, dataErrors: errorsMap };
+  }
+}
+
+function validate(
+  jsonSchema: JSONSchema,
+  data: unknown
+): { valid: boolean; errors?: Record<string, any> } {
+  const validator = ajv.compile(jsonSchema);
+  const valid = validator(data) as boolean;
+  const errors = validator.errors?.map((e) => e);
+  return { valid, errors: errors?.map(toErrorDetail) };
+}
+
+function toErrorDetail(error: ErrorObject) {
+  return {
+    path: error.instancePath.substring(1).replace(/\//g, '.') || '(root)',
+    message: error.message!,
+  };
+}
+
+async function getAction(
+  manager: RuntimeManager,
+  actionRef: string
+): Promise<Action | undefined> {
+  const actions = await manager.listActions();
```

Comment:
just FYI, but this is an expensive call... it's OK depending on application, but in this case seems fine.

Reply:
Would you rather have a `getAction` call here instead?